crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -0,0 +1,808 @@
+        engine: TestBeaconConsensusEngine,
+    ) -> oneshot::Receiver<Result<(), BeaconEngineError>> {
+        let (tx, rx) = oneshot::channel();
+        tokio::spawn(async move {
+            let result = engine.await;
+            tx.send(result).expect("failed to forward consensus engine result");
+        });
+        rx
+    }
+
+    // Pipeline error is propagated.
+    #[tokio::test]
+    async fn pipeline_error_is_propagated() {
+        let chain_spec = Arc::new(
+            ChainSpecBuilder::default()
+                .chain(MAINNET.chain)
+                .genesis(MAINNET.genesis.clone())
+                .paris_activated()
+                .build(),
+        );
+        let (consensus_engine, env) = setup_consensus_engine(
+            chain_spec,
+            VecDeque::from([Err(StageError::ChannelClosed)]),
+            Vec::default(),
+        );
+        let rx = spawn_consensus_engine(consensus_engine);
+
+        let _ = env.send_forkchoice_updated(ForkchoiceState {
+            head_block_hash: H256::random(),
+            ..Default::default()
+        });
+        assert_matches!(
+            rx.await,
+            Ok(Err(BeaconEngineError::Pipeline(PipelineError::Stage(StageError::ChannelClosed))))
+        );
+    }
+
+    // Test that the consensus engine is idle until first forkchoice updated is received.
+    #[tokio::test]
+    async fn is_idle_until_forkchoice_is_set() {
+        let chain_spec = Arc::new(
+            ChainSpecBuilder::default()
+                .chain(MAINNET.chain)
+                .genesis(MAINNET.genesis.clone())
+                .paris_activated()
+                .build(),
+        );
+        let (consensus_engine, env) = setup_consensus_engine(
+            chain_spec,
+            VecDeque::from([Err(StageError::ChannelClosed)]),
+            Vec::default(),
+        );
+        let mut rx = spawn_consensus_engine(consensus_engine);
+
+        // consensus engine is idle
+        std::thread::sleep(Duration::from_millis(100));
+        assert_matches!(rx.try_recv(), Err(TryRecvError::Empty));
+
+        // consensus engine is still idle
+        let _ = env.send_new_payload(SealedBlock::default().into());
+        assert_matches!(rx.try_recv(), Err(TryRecvError::Empty));
+
+        // consensus engine receives a forkchoice state and triggers the pipeline
+        let _ = env.send_forkchoice_updated(ForkchoiceState {
+            head_block_hash: H256::random(),
+            ..Default::default()
+        });
+        assert_matches!(
+            rx.await,
+            Ok(Err(BeaconEngineError::Pipeline(PipelineError::Stage(StageError::ChannelClosed))))
+        );
+    }
+
+    // Test that the consensus engine runs the pipeline again if the tree cannot be restored.
+    // The consensus engine will propagate the second result (error) only if it runs the pipeline
+    // for the second time.
+    #[tokio::test]
+    async fn runs_pipeline_again_if_tree_not_restored() {
```

Comment:
Why is the tree not restored in this test? What would make it succeed which is not present here?

Reply:
the finalized block hash is not in the db, so the blockchain tree cannot be restored